### DISCUSSION TITLE
Why does READ-FILE and READ-LINE return values not the same?

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#160] [Q&A] [answered]
`READ-FILE ( caddr u1 fid -- u2 ior )` while `READ-LINE ( caddr u1 fid -- u2 flag ior )` has an extra redundant `flag`.  Surely in the case of `READ-LINE` `ior == 0` is sufficient to indicate success as is done else where; the extra `flag` return value seems pointless.

**Comment by alextangent** [comment, marked as answer]
As specified in the standard, it is there to ensure that we can tell the difference between end-of-file and a zero length line. A single return value is not sufficient, as is explained in the annex A; https://forth-standard.org/standard/rationale#rat:file:READ-LINE. 

 

READ-FILE does not require it since there is no concept of a lines, just a requested and returned length.

 

 

From: Anthony Howe ***@***.***> 
Sent: Thursday, August 8, 2024 3:33 PM
To: ForthHub/discussion ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [ForthHub/discussion] Why does READ-FILE and READ-LINE return values not the same? (Discussion #160)

 

READ-FILE ( caddr u1 fid -- u2 ior ) while READ-LINE ( caddr u1 fid -- u2 flag ior ) has an extra redundant flag. Surely in the case of READ-LINE ior == 0 is sufficient to indicate success as is done else where; the extra flag return value seems pointless.

**Reply by ruv** [reply]
> it is there to ensure that we can tell the difference between end-of-file and a zero length line.

BTW, an alternative way could be to store two null characters after _u_ characters in the buffer if no line terminator was read. If a line terminator was read, these two characters are not both null characters.  But checking the end-of-file condition becomes a little more difficult.
